### PR TITLE
self-hosted: hook up Zig AST to ZIR

### PR DESCRIPTION
**Pull request by andrewrk**
* Introduce the concept of anonymous Decls
 * Primitive Hello, World with inline asm works
 * There is still an unsolved problem of how to manage ZIR instructions
   memory when generating from AST. Currently it leaks.
 * Decl references and module updates not handled yet.

This example works:

```zig
export fn _start() noreturn {
    asm volatile ("syscall"
        :
        : [number] "{rax}" (1),
          [arg1] "{rdi}" (1),
          [arg2] "{rsi}" (@ptrToInt("Hello, world!\n")),
          [arg3] "{rdx}" (14)
        : "rcx", "r11", "memory"
    );
    asm volatile ("syscall"
        :
        : [number] "{rax}" (231),
          [arg1] "{rdi}" (0)
        : "rcx", "r11", "memory"
    );
    unreachable;
}
```

### CODE REVIEW

File: lib/std/zig.zig [hunk 1]
```diff
@@ -9,6 +11,21 @@ pub const ast = @import("zig/ast.zig");
 pub const system = @import("zig/system.zig");
 pub const CrossTarget = @import("zig/cross_target.zig").CrossTarget;
 
+pub const SrcHash = [16]u8;
+
+/// If the source is small enough, it is used directly as the hash.
```
Comment (daurnimator) [inline lib/std/zig.zig:16]: this seems like a recipe for collisions. blake3 should be fast enough that we should play it safe...

Comment (pixelherodev) [reply lib/std/zig.zig:16]: Ahh, because a real hash will collide with the source?

**Comment by andrewrk** [comment]
There is still some work to do to finish this branch and get it ready for merging, but I have implemented Decl lookup, and now this test case works:

```zig
export fn _start() noreturn {
    print();
    print();
    print();
    print();
    print();

    exit();
}

fn print() void {
    asm volatile ("syscall"
        :
        : [number] "{rax}" (1),
          [arg1] "{rdi}" (1),
          [arg2] "{rsi}" (@ptrToInt("Hello, world!\n")),
          [arg3] "{rdx}" (14)
        : "rcx", "r11", "memory"
    );
    return;
}

fn exit() noreturn {
    asm volatile ("syscall"
        :
        : [number] "{rax}" (231),
          [arg1] "{rdi}" (0)
        : "rcx", "r11", "memory"
    );
    unreachable;
}
```

Just for fun, I tried putting 1 million `print()` calls in there, to see how it performed:

https://asciinema.org/a/LvkNeNjREHROBEEFvsrg1FxWi

It's a silly but fun metric. On my laptop, it compiled in 0.79 seconds, using only 610KB resident memory to do the job.

**Comment by pixelherodev** [comment]
> It's a silly but fun metric. On my laptop, it compiled in 0.66 seconds, using 471 MiB peak resident memory to do the job. An equivalent .c source code took 7x as long in Clang and used 3x as much peak memory.

Mind giving a GCC metric? ;)

**Comment by andrewrk** [comment]
> Mind giving a GCC metric? ;)

Sure. Compared to the equivalent Zig code, GCC 9.2.0 took 18x as long and used 11x more memory.

**Comment by andrewrk** [comment]
* [ ] figure out memory management / lifecycle for anonymous Decls
 * [ ] hook up incremental compilation
 * [ ] fix regressions to ZIR test cases
 * [ ] add additional test cases for Zig AST source code

**Comment by andrewrk** [comment]
Updated demo with incremental compilation wired up: https://asciinema.org/a/LcwgiZzoIPKAfb6Jr8iDOqPJ9

First build:
image

Updating the message. you can see nearly all the time is spent in parsing the AST
image

Now it's just those 2 remaining check boxes, then this should be good to merge.

**Pull request merged**